File: backend/schemas/experience.py
from pydantic import BaseModel, UUID4
from typing import List, Optional
from datetime import date, datetime


class SkillBase(BaseModel):
    name: str


class SkillCreate(SkillBase):
    pass


class Skill(SkillBase):
    id: UUID4
    
    class Config:
        from_attributes = True


class ExperienceBase(BaseModel):
    company_name: str
    title: str
    location: Optional[str] = None
    start_date: date
    end_date: Optional[date] = None
    is_current: bool = False
    description: str


class ExperienceCreate(ExperienceBase):
    pass


class ExperienceUpdate(BaseModel):
    company_name: Optional[str] = None
    title: Optional[str] = None
    location: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_current: Optional[bool] = None
    description: Optional[str] = None


class Experience(ExperienceBase):
    id: UUID4
    user_id: UUID4
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True 

File: frontend/app/(cover-letter-app)/api/experience.ts
import { getAuthFromStorage } from '@/lib/sessionStorage';
import { authenticatedRequest } from '@/lib/apiClient';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api';

// Define types based on our backend schema
export interface Skill {
  id: string;
  name: string;
}

export interface ExperienceBase {
  company_name: string;
  title: string;
  location?: string;
  start_date: string; // ISO format date string
  end_date?: string | null; // ISO format date string, can be null
  is_current: boolean;
  description: string;
}

export interface ExperienceCreate extends ExperienceBase {
  // Remove skills field since it's not implemented in backend
}

export interface ExperienceUpdate {
  company_name?: string;
  title?: string;
  location?: string;
  start_date?: string;
  end_date?: string | null;
  is_current?: boolean;
  description?: string;
  // Remove skills field
}

export interface Experience extends ExperienceBase {
  id: string; // UUID as string
  user_id: string; // UUID as string
  // Remove skills field
  created_at?: string; // ISO format datetime string
  updated_at?: string; // ISO format datetime string
}

// Create a new experience
export const createExperience = async (experience: ExperienceCreate): Promise<Experience> => {
  try {
    return await authenticatedRequest<Experience>('experiences', {
      method: 'POST',
      body: JSON.stringify(experience)
    });
  } catch (error) {
    console.error("Error creating experience:", error);
    throw error;
  }
};

// Get all experiences for the current user
export const getUserExperiences = async (): Promise<Experience[]> => {
  try {
    return await authenticatedRequest<Experience[]>('experiences', {
      method: 'GET'
    });
  } catch (error) {
    console.error("Error fetching experiences:", error);
    throw error;
  }
};

// Get a specific experience by ID
export const getExperience = async (experienceId: string): Promise<Experience> => {
  try {
    return await authenticatedRequest<Experience>(`experiences/${experienceId}`, {
      method: 'GET'
    });
  } catch (error) {
    console.error(`Error fetching experience ${experienceId}:`, error);
    throw error;
  }
};

// Update an existing experience
export const updateExperience = async (experienceId: string, update: ExperienceUpdate): Promise<Experience> => {
  try {
    return await authenticatedRequest<Experience>(`experiences/${experienceId}`, {
      method: 'PUT',
      body: JSON.stringify(update)
    });
  } catch (error) {
    console.error(`Error updating experience ${experienceId}:`, error);
    throw error;
  }
};

// Delete an experience
export const deleteExperience = async (experienceId: string): Promise<void> => {
  try {
    await authenticatedRequest(`experiences/${experienceId}`, {
      method: 'DELETE'
    });
  } catch (error) {
    console.error(`Error deleting experience ${experienceId}:`, error);
    throw error;
  }
};
